OuterTypeNumberCheck reports false positive on JEP 512 compact source files

Parent Issue #19971

I have read check documentation: https://checkstyle.org/checks/sizes/outertypenumber.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > config.xml <<'EOF'
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="OuterTypeNumber"/>
  </module>
</module>
EOF

vivek@Viveks-MacBook-Air checkstyle % cat > InputOuterTypeNumberCompact.java <<'EOF'
void main() { }

class A { }
class B { }
EOF

vivek@Viveks-MacBook-Air checkstyle % RUN_LOCALE="-Duser.language=en -Duser.country=US"
java $RUN_LOCALE -jar target/checkstyle-13.6.0-SNAPSHOT-all.jar -c config.xml InputOuterTypeNumberCompact.java

Starting audit...
[ERROR] InputOuterTypeNumberCompact.java:1:1: Outer types defined is 2 (max allowed is 1). [OuterTypeNumber]
Audit done.
Checkstyle ends with 1 errors.
```

---

**Describe what you expect in detail.**

OuterTypeNumberCheck must not flag the compact file. In JEP 512 semantics, `class A` and `class B` are nested types of the implicit class that the compiler synthesises for a compact source file. They are members of the implicit class, not outer types. The compact file contains exactly one outer type (the implicit class), so the check should report zero violations.

The check currently counts every `CLASS_DEF` (and the other type-def tokens) child of the root and increments `currentOuterTypeCount` once per child. Under the new grammar the root is `COMPACT_COMPILATION_UNIT` and its `CLASS_DEF` children are member declarations, but the check treats them the same as `COMPILATION_UNIT` children.

Expected on the compact input: zero violations.

**Same code inside an ordinary class:**

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > InputOuterTypeNumberOrdinary.java <<'EOF'
public class InputOuterTypeNumberOrdinary {
    public void main() { }

    class A { }
    class B { }
}
EOF

vivek@Viveks-MacBook-Air checkstyle % java $RUN_LOCALE -jar target/checkstyle-13.6.0-SNAPSHOT-all.jar -c config.xml InputOuterTypeNumberOrdinary.java

Starting audit...
Audit done.
```

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/OuterTypeNumberCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/OuterTypeNumberCheck.java
@@ -51,6 +51,13 @@ public class OuterTypeNumberCheck extends AbstractCheck {
     /** Tracks the number of outer types found. */
     private int outerNum;
 
+    /**
+     * Whether the file is a JEP 512 compact source file. All its type
+     * declarations are members of the implicit class, so it always has exactly
+     * one outer type and the limit can never be exceeded.
+     */
+    private boolean isCompactSourceFile;
+
     @Override
     public int[] getDefaultTokens() {
         return getRequiredTokens();
@@ -76,11 +83,13 @@ public class OuterTypeNumberCheck extends AbstractCheck {
     public void beginTree(DetailAST ast) {
         currentDepth = 0;
         outerNum = 0;
+        isCompactSourceFile = ast != null
+                && ast.getType() == TokenTypes.COMPACT_COMPILATION_UNIT;
     }
 
     @Override
     public void finishTree(DetailAST ast) {
-        if (max < outerNum) {
+        if (!isCompactSourceFile && max < outerNum) {
             log(ast, MSG_KEY, outerNum, max);
         }
     }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/sizes/OuterTypeNumberCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/sizes/OuterTypeNumberCheckTest.java
@@ -110,6 +110,13 @@ public class OuterTypeNumberCheckTest extends AbstractModuleTestSupport {
                 getPath("InputOuterTypeNumberRecords.java"), expected);
     }
 
+    @Test
+    public void testCompactSourceFile() throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputOuterTypeNumberCompactSourceFile.java"), expected);
+    }
+
     /**
      * Checks if the private field {@code currentDepth} and {@code outerNum} is
      * properly cleared during the start of processing the next file in the

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/sizes/outertypenumber/InputOuterTypeNumberCompactSourceFile.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/sizes/outertypenumber/InputOuterTypeNumberCompactSourceFile.java
@@ -1,0 +1,36 @@
+/*
+OuterTypeNumber
+max = (default)1
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+// Every type declared below is a member of the implicit class synthesized for
+// this compact source file, not an outer type. The only outer type is the
+// implicit class itself, so the check must stay silent even though several
+// types of every kind, plus nested types, are declared.
+
+void main() {
+    System.out.println("Hello!");
+}
+
+class A {
+    class Inner {
+    }
+}
+
+interface B {
+}
+
+enum C {
+    ONE,
+    TWO
+}
+
+record D(int x) {
+}
+
+@interface E {
+}

